### PR TITLE
#170766088 Logout Functionality

### PR DESCRIPTION
**Pull request by key-joshua**
## WHAT DOES THIS PR DO?

- I have created a logout endpoint.
- I have created a table to store a valid token.

## DESCRIPTION OF TASK TO BE COMPLETED?

- During the completion of this work I have created logout endpoint which will be used to logout from application
  and also I have created a table which will store a valid token.

####  The way this will be tested:

- Users should pull from this branch or clone this repository into his/her machine.
- Users after clone this repository should run ```npm install``` command to install all required tools.
- Users after install all required tools should open SQL Shell (psql) and run ```CREATE DATABASE barefoot```.
- Users after creating this Database should run ```createAndInsertIntoValidTokensTable``` script to create and insert data into a table.
- Users after create and insert data should run ```npm run server``` command to start the server and should have to use this URL when tested local ```localhost:8080/api/v-unknown/users/logout``` in postman.

#### When a request runs perfect will get a response in this format below:
RESPONSE:
----------------
status code (200 Ok )
message (string)
```
{
	"status":  200,
	"message":   " Hey username!! you are logged out successfully"
	
}
```

## ANY BACKGROUND CONTEXT YOU  WANT TO PROVIDE?

- The context was all about NODEJS with LIBRARIES.

## WHAT ARE THE RELEVANT PIVOTAL TRACKER STORIES?
- [feature: Logout Functionality](https://www.pivotaltracker.com/story/show/170766088)

#### For Swagger
logouts

#### For Postman
logout

### CODE REVIEW

**Review by Nigorjeanluc — DISMISSED** [review]

**Review by jajajaden01 — DISMISSED** [review]

**Review by karengir — DISMISSED** [review]

**Review by Nigorjeanluc — DISMISSED** [review]

**Review by izabayogit — APPROVED** [review]

**Review by jajajaden01 — APPROVED** [review]

File: server/controllers/authUserController.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+class AuthUsers {
+  async logout(req, res) {
+    try {
+      if (req.headers.authorization) {
+        await importDeleteQuery.deleteValidToken(req.headers.authorization);
+        return res
+          .status(200)
+          .json({ status: 200, message: ` Hey Joshua !! you are logged out successfully` });
+      }
+      return res
```
Comment (luc-tuyishime) [inline server/controllers/authUserController.js]: Why returning two times??

File: server/models/migrations/insertValidTokenTableQuery.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import databaseConnection from '../../config/dbConnectWithSequelize';
+import validTokenTable from './createValidTokenTableQuery';
+
+const valitoken1 = `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwiY2F0ZWdvcnkiOiJhZG1pbiIsImZpcnN0bmFtZSI6Im1pbmFuaSIsImxhc3RuYW1lIjoiam9zaHVhIiwidXNlcm5hbWUiOiJrZXktam9zaHVhIiwiZW1haWwiOiJrLmpvc2h1YTg1NUBnbWFpbC5jb20iLCJwaG9uZW51bWJlciI6IjA3ODk2MTk0NDIiLCJwYXNzd29yZCI6IiQyYSQxMiQzVDN3RG1QZU5CYVplOXVGZm0yM3plZ3VVT29Sd1d6cDVuMUU3LnJUWnNWcVp1aXBLN1NvbSIsImNyZWF0ZWRkYXRlIjoiMjAxOS0xMS0wNFQxNjowMjoyMi4zMTVaIiwibW9kaWZpZWRkYXRlIjoiIG5vbmUiLCJpYXQiOjE1ODAyNzg5MjgsImV4cCI6MTY2NjY3ODkyOH0.GBTZe8FjA7dvw3NWlPfXVQxBEHFlap6798A71C61fs0`;
+module.exports = databaseConnection.sync({ logging: console.log, force: true }).then(() => {
+  validTokenTable.create({ token: valitoken1 });
+});
```
Comment (luc-tuyishime) [inline server/models/migrations/insertValidTokenTableQuery.js]: What is the use of this file??

File: server/controllers/authUserController.js (hunk 1 above)
Comment (key-joshua) [reply server/controllers/authUserController.js]: For the first will be returned in the case token detected from a header and at the second will be returned when there is no token detected from a header.

File: server/models/migrations/insertValidTokenTableQuery.js (hunk 2 above)
Comment (key-joshua) [reply server/models/migrations/insertValidTokenTableQuery.js]: This file will be used to delete existing data and insert data again into a table after table being created into a database and will help my test running

File: server/controllers/authUserController.js (hunk 1 above)
Comment (key-joshua) [reply server/controllers/authUserController.js]: I corrected this mistake after analysing the useless algorithm I used before, now everything looks great

File: server/tests/logoutTest.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  //       expect(response).to.have.status([200]);
+  //       expect(response.body).to.be.a('object');
+  //       expect(response.body).to.have.property('status');
+  //       expect(response.body.status).to.be.equal(200);
+  //       expect(response.body).to.have.property('message');
+  //       expect(response.body.message).to.be.a('string');
+  //       done(error);
+  //     });
+  // });
+});
```
Comment (luc-tuyishime) [inline server/tests/logoutTest.js:55]: Why do you push commented codes??

Comment (key-joshua) [reply server/tests/logoutTest.js:55]: Ooohhh... let correct this.

File: .eslintrc [hunk 4]
```diff
@@ -7,30 +7,45 @@
...
+      "error",
+      {
+        "requireReturn": true,
+        "requireReturnType": true,
+        "requireParamDescription": false,
+        "requireReturnDescription": true
+      }
+    ],
+    "require-jsdoc": [
+      0,
```
Comment (Nigorjeanluc) [inline .eslintrc]: Why are you changing these rules?

File: server/index.js [hunk 5]
```diff
@@ -1,21 +1,23 @@
 import express from 'express';
 import swaggerUi from 'swagger-ui-express';
 import apiDocumentation from '../swagger.json';
+import importAuthUserRoute from './routes/authUserRoute';
 
 const app = express();
+const basePath = '/api/v-unknown';
```
Comment (Nigorjeanluc) [inline server/index.js]: I think '/api' would suffice.  '/api' here is enough

File: .eslintrc (hunk 4 above)
Comment (Nigorjeanluc) [inline .eslintrc]: This rule is necessary to help teammates understand what your functions do.

File: .eslintrc [hunk 6]
```diff
@@ -7,30 +7,45 @@
...
+    "prefer-template": 0,
     "one-var-declaration-per-line": 0,
+    "import/newline-after-import": 0,
+    "class-methods-use-this": 0,
     "new-cap": 0,
     "consistent-return": 0,
     "no-param-reassign": 0,
     "comma-dangle": 0,
-    "arrow-parens": ["error", "always"],
+    "arrow-parens": [0, "error", "always"],
```
Comment (Nigorjeanluc) [inline .eslintrc]: I think we must adhere to the same rules.

**Review by Nigorjeanluc — CHANGES_REQUESTED** [review]

File: .eslintrc (hunk 6 above)
Comment (key-joshua) [reply .eslintrc]: Ooooh... I forgot to undo the changes

**Review by key-joshua — COMMENTED** [review]
@Nigorjeanluc Thank you for your feedback

File: .eslintrc (hunk 4 above)
Comment (key-joshua) [reply .eslintrc]: Yeah Exactly

Comment (key-joshua) [reply .eslintrc]: sometimes I made changes when am working then I forget to undo

File: server/index.js (hunk 5 above)
Comment (key-joshua) [reply server/index.js]: Ooohhkk... I  was guessing maybe we will add a version of our application that why I added Version Unknown but it fine.

**Review by Nigorjeanluc — APPROVED** [review]

**Comment by KKDoddy** [comment]
We should check the swagger documentation first and confirm that it is working fine

**Pull request merged**